File: frontend/lib/models/declaration.dart
class Declaration {
  final int id;
  final int clientId;
  final int mois;
  final int annee;
  final double chiffreAffaires;
  final double totalVentes;
  final double totalAchats;
  final int nombreEmployes;
  final String? observations;
  final String statut;
  final String? commentaireAdmin;
  final DateTime dateCreation;
  final DateTime? dateSoumission;
  final DateTime? dateValidation;
  final int? traiteParId;

  const Declaration({
    required this.id,
    required this.clientId,
    required this.mois,
    required this.annee,
    required this.chiffreAffaires,
    required this.totalVentes,
    required this.totalAchats,
    required this.nombreEmployes,
    this.observations,
    required this.statut,
    this.commentaireAdmin,
    required this.dateCreation,
    this.dateSoumission,
    this.dateValidation,
    this.traiteParId,
  });

  factory Declaration.fromJson(Map<String, dynamic> json) {
    return Declaration(
      id: (json['id'] as num).toInt(),
      clientId: (json['client_id'] as num).toInt(),
      mois: (json['mois'] as num).toInt(),
      annee: (json['annee'] as num).toInt(),
      chiffreAffaires:
          double.parse(json['chiffre_affaires'].toString()),
      totalVentes:
          double.parse(json['total_ventes'].toString()),
      totalAchats:
          double.parse(json['total_achats'].toString()),
      nombreEmployes:
          (json['nombre_employes'] as num).toInt(),
      observations: json['observations']?.toString(),
      statut: json['statut']?.toString() ?? 'BROUILLON',
      commentaireAdmin:
          json['commentaire_admin']?.toString(),
      dateCreation:
          DateTime.parse(json['date_creation'].toString()),
      dateSoumission: json['date_soumission'] != null
          ? DateTime.parse(json['date_soumission'].toString())
          : null,
      dateValidation: json['date_validation'] != null
          ? DateTime.parse(json['date_validation'].toString())
          : null,
      traiteParId: json['traite_par_id'] != null
          ? (json['traite_par_id'] as num).toInt()
          : null,
    );
  }

  String get moisLabel {
    const labels = [
      '',
      'Janvier',
      'Février',
      'Mars',
      'Avril',
      'Mai',
      'Juin',
      'Juillet',
      'Août',
      'Septembre',
      'Octobre',
      'Novembre',
      'Décembre',
    ];

    return labels[mois];
  }

  String get periode => '$moisLabel $annee';

  bool get estBrouillon => statut == 'BROUILLON';
  bool get estSoumise => statut == 'SOUMISE';
  bool get enVerification => statut == 'EN_VERIFICATION';
  bool get estValidee => statut == 'VALIDEE';
  bool get aCorriger => statut == 'A_CORRIGER';
  bool get estRejetee => statut == 'REJETEE';

  String get statutLabel {
    switch (statut) {
      case 'BROUILLON':
        return 'Brouillon';
      case 'SOUMISE':
        return 'Soumise';
      case 'EN_VERIFICATION':
        return 'En vérification';
      case 'VALIDEE':
        return 'Validée';
      case 'A_CORRIGER':
        return 'À corriger';
      case 'REJETEE':
        return 'Rejetée';
      default:
        return statut;
    }
  }
}

File: backend/app/models/declaration.py
from datetime import datetime

from sqlalchemy import (
    Column,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import relationship

from app.core.database import Base


class Declaration(Base):
    __tablename__ = "declarations"

    id = Column(Integer, primary_key=True, index=True)

    # Client concerné
    client_id = Column(
        Integer,
        ForeignKey("clients.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Période de déclaration
    mois = Column(Integer, nullable=False)
    annee = Column(Integer, nullable=False)

    # Données déclarées
    chiffre_affaires = Column(
        Numeric(15, 2),
        nullable=False,
        default=0,
    )

    total_ventes = Column(
        Numeric(15, 2),
        nullable=False,
        default=0,
    )

    total_achats = Column(
        Numeric(15, 2),
        nullable=False,
        default=0,
    )

    nombre_employes = Column(
        Integer,
        nullable=False,
        default=0,
    )

    observations = Column(Text, nullable=True)

    # BROUILLON / SOUMISE / EN_VERIFICATION / VALIDEE / A_CORRIGER / REJETEE
    statut = Column(
        String(30),
        nullable=False,
        default="BROUILLON",
        index=True,
    )

    # Commentaire du fiscaliste/admin lors d'une correction ou d'un rejet
    commentaire_admin = Column(Text, nullable=True)

    date_creation = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
    )

    date_soumission = Column(
        DateTime,
        nullable=True,
    )

    date_validation = Column(
        DateTime,
        nullable=True,
    )

    # Utilisateur ayant traité la déclaration
    traite_par_id = Column(
        Integer,
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    client = relationship(
        "Client",
        back_populates="declarations",
    )

    traite_par = relationship(
        "User",
        foreign_keys=[traite_par_id],
    )

    documents = relationship(
        "Document",
        back_populates="declaration",
        cascade="all, delete-orphan",
    )

    __table_args__ = (
        UniqueConstraint(
            "client_id",
            "mois",
            "annee",
            name="uq_declaration_client_mois_annee",
        ),
    )
